### PR TITLE
New data schema with 2 million books (+ new filtering setup)

### PR DESCRIPTION
**Pull request by leerob**
### Overview

This PR moves from the previous 50,000 book database from Goodreads (with higher quality photos) to a much, much larger dataset with 2 million books and 800,000 authors. Adding more data, which is more realistic with a very large application with books, cars, movies, or other massive datasets, means you need to construct the application differently.

For example, I could no longer show the authors panel on the left hand side. It's not really useful anymore with that many authors. If you want to search for an author, you would instead use the normal search bar. Similarly, when you have that many rows in the database, doing the total count of the table in the same function you're making the search query doesn't make sense. You need to `limit` the search.

Instead, I created a new sidebar that allows you to filter the books in different ways. By the number of pages, year published, average star rating, and more. This is much more useful, and is a better example of using URL state. In the process of building this, I refactored out the data layer logic.

### Seeding Data

Getting all of that data into Postgres takes a nice script! I wrote some utils that were shared to read the JSON files as a stream (because they're massive), go over them line by line and save checkpoints (so you can resume), and support batching. It even has a nice little progress `console.log` as we go.

I first ran the authors, then the books, since the books also create the relations to the authors. Finally, I added another script for generating a [thumbhash](https://evanw.github.io/thumbhash/) for every photo. This gives it the nice blur-up placeholder effect, even with remote URLs (from Goodreads URLs). Next.js can generate blur-up placeholders for you when the file is on the file system, but when it's remote, you have to store the base64 generated hash into the database.

### Misc

- Since the new photos are lower quality, I made the grid bigger
- Clicking into a book now retains the URL params (so the filters persist across transitions)
- When querying books to show, we only every show books with images (not a perfect dataset)
- The back link navigating from a books page can now be opened in a new tab (actual link)

### CODE REVIEW

File: lib/db/queries.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  return and(
+    not(isNull(books.image_url)),
+    sql`${books.image_url} != ${EMPTY_IMAGE_URL}`
+  );
+};
+
+const isbnFilter = (isbn?: string) => {
+  if (isbn) {
+    const isbnArray = isbn.split(',').map((id) => id.trim());
+    return sql`books.isbn IN (${sql.join(
```
Comment (AlexBlokh) [inline lib/db/queries.ts:65]: maybe we should use `inArray`?
please let me know if that didn't work for some reason

File: lib/db/schema.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  {
+    bookId: integer('book_id')
+      .notNull()
+      .references(() => books.id),
+    authorId: text('author_id')
+      .notNull()
+      .references(() => authors.id),
+  },
+  (t) => ({
+    pk: primaryKey({ columns: [t.bookId, t.authorId] }),
```
Comment (AlexBlokh) [inline lib/db/schema.ts:81]: I'd suggest to add indexes on both fields for `joins` performance

File: lib/db/queries.ts [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+      text_reviews_count: books.text_reviews_count,
+      ratings_count: books.ratings_count,
+      average_rating: books.average_rating,
+      series: books.series,
+      createdAt: books.createdAt,
+      authors: sql<string[]>`array_agg(${authors.name})`,
+      thumbhash: books.thumbhash,
+    })
+    .from(books)
+    .leftJoin(bookToAuthor, eq(books.id, bookToAuthor.bookId))
```
Comment (AlexBlokh) [inline lib/db/queries.ts:152]: You could omit indexes I've mentioned below if you'd query from `bookToAuthor` table first with joins on `books.id` and `authors.id`

**Comment by leerob** [comment]
I'm going to merge forward and then let's work on a PR together to optimize drizzle from here 😄

**Pull request merged**